apps/prairielearn/src/lib/editors.ts

```diff
@@ -231,16 +247,32 @@ export abstract class Editor {
           );
         };
 
+        let sharingConfigurationValid = true;
+        let courseData;
         try {
           await writeAndCommitChanges();
 
+          await cleanAndResetRepository(this.course, 'HEAD', gitEnv, job);
```

Comment:
This is new since the last time I looked at this PR. If I'm following the flow right, this means that on the happy path (we can push without having to integrate remote changes first), we actually do this clean-and-reset sequence twice, which adds about 150ms of latency for larger courses.

I think you can get away with removing the final `cleanAndResetRepository` (currently line 310) and adding a call after the call to `writeAndCommitChanges()` in the "Failed to push changes to remote git repository" catch block. Does that sound correct to you?

Reply:
That's right, I had to add this because in some edit operation, a file or folder sitting around and got picked up by the first round of parsing the course data, which wasn't actually commited to the repo. This was caught by `courseEditor.test.ts` so good job to whoever wrote those tests!

Yes, I think your proposed solution works, I'll do that.

Reply:
actually, I think it needs to go in the catch block of the other try catch block (there are two nested try catch blocks here. I've put it there, let me know what you think.

Reply:
Actually I've change my mind. I'm going to put it in the catch block you mentioned, but also in the `if` statement that checks for the sharing config being valid. That should have the same behaviour but be a little more clear/readable.

Reply:
I was in the middle of typing out exactly that suggestion 🙂

Reply:
Glad we're on the same page :)